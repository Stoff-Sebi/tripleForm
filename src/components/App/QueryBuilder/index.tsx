import React from "react";
interface props {
  queryStart: string;
  query: string;
  paramDelimiter: string;
}

const QueryBuilder: React.FC<props> = ({
  queryStart,
  query,
  paramDelimiter
}) => {
  return (
    <div className={`container`}>
      <br></br>
      <h4>Query Build Helper</h4>
      <p>
        (set lifecycle to "deploy" to deactivate builder) <br></br>
        <br></br>Query-Start: <br></br>
        {queryStart}
      </p>{" "}
      <hr></hr>
      <p>
        Decoded: <br></br> {query}
      </p>{" "}
      <hr></hr>
      <p>
        Encoded:<br></br> {encodeURIComponent(query)}
      </p>{" "}
      <hr></hr>
      <p>
        Given parameter-delimiter:<br></br> {paramDelimiter}
      </p>{" "}
    </div>
  );
};

export default QueryBuilder;
