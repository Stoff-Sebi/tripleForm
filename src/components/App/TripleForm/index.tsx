import React from "react";
import styles from "./styles.module.css"; //import css module here -> this variable can then be used in classname.
import ResponsiveForm from "../../common/ResponsiveForm"
import {QueryInput, QuerySelectInput, SelectValue} from "../../../@types/types";

interface props {
  queryStart?: string;
  parameterDelimiter?: string;
  parameters?: QueryInput[];
  encode?: boolean;
}

const TripleForm: React.FC<props> = ({queryStart="https://glossa.uni-graz.at/archive/objects/query:cantus.fulltext/methods/sdef:Query/get?params=", parameterDelimiter = ";", parameters = [
{
  label: "Cantus Volltextsuche",
  value:"",
  type: "text",
  id:"cantus_01_test",
  placeHolder:"z.b. Salzburg",
  parameter: "$5|"
},
{
  label:"test02",
  value:"bla",
  type: "text",
  id:"123",
  placeHolder:"bimbi",
  parameter: "$2|"
},
{
  label:"test03",
  value:"blasdas",
  type: "text",
  id:"1233333sssss",
  placeHolder:"bimbiasdds",
  parameter: "$3|"
},
{
  label:"test04",
  value:[{label:"firstVal", value:"1", _selected:true},{label:"secondVal", value:"2", _selected:false}],
  type: "select",
  id:"1233333as",
  placeHolder:"bimbiasdds",
  parameter: "$8|"
},
{
  label:"test05",
  value:[{label:"Vierzehn", value:"14", _selected:true},{label:"Hupfburg", value:"Hupfburg", _selected:false}],
  type: "select",
  id:"1233333as",
  placeHolder:"bimbiasdds", 
  parameter: "$9|"
},
{
  label:"test05",
  value:[{label:"Vierzehn", value:"14", _selected:true},{label:"Hupfburg", value:"Hupfburg", _selected:false}],
  type: "autocomplete",
  id:"1233333as",
  placeHolder:"bimbiasdds", 
  parameter: "$10|"
}

]}) => {

  const [query, setQuery] = React.useState<""| string>("");
  const [queryInputs, setInputs] = React.useState<QueryInput[]>(parameters);

  const handleSearch = (btnClickEvent: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    btnClickEvent.preventDefault(); //prevents default submit action on form.
    if(!query)return alert("wählen Sie einen gültigen Wert für die Suche aus.");
    let inputErrFlag: boolean = false;
    queryInputs.forEach(queryInput=>{
      if(!queryInput.value){
        inputErrFlag = true;
      }
    })
    if(inputErrFlag)return alert("Bitte wählen Sie für alle Suchfelder einen gültigen Wert aus.");
    let url = queryStart + encodeURIComponent(query);
    window.location.href = url;
  }

  React.useEffect(()=>{
    console.log(queryInputs);
    if(!queryInputs) return;
    console.log("Inputs changed!", queryInputs);

    let query = "";
    queryInputs.forEach(queryInput => {
      if(query==="undefined")query = "";

      //check which type queryInput has
      if(queryInput.type==="text"){
        query += `${queryInput.parameter}${queryInput.value}${queryInputs.length > 1 ? parameterDelimiter : ''}`;
      }

      //check if type is queryInput
      if(queryInput.type==="select"){
        //if array
        if(Array.isArray(queryInput.value)){
          (queryInput.value as SelectValue[]).forEach((inputObj) => {
            //if _selected property set to true
            if(inputObj._selected === true)query += `${queryInput.parameter}${inputObj.value}${queryInputs.length > 1 ? parameterDelimiter : ''}`;
          })
        } else {
          throw new TypeError(`Encountered a not array type inside an queryInput marked as 'select'. Input's label is: ${queryInput.label}`);
        }
      }
       
    });
    
    if(query==="undefined")return setQuery("");
    setQuery(query)
  }, [queryInputs]);

  return (
    <>
    <p>{query}</p>
    <ResponsiveForm
      inputFields={queryInputs}
      setInputFields={setInputs}
      handleSearch={handleSearch}
    ></ResponsiveForm>
    </>
  );
};

export default TripleForm;
