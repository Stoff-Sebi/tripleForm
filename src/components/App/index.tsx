import React from "react";
import GamsWidget from "../common/GamsWidget";
import TripleForm from "./TripleForm";
import WidgetDefProvider from "../common/GamsWidget/WidgetDefProvider"
import DataProvider from "../common/GamsWidget/DataProvider"

const TripleFormWidget: React.FC = () => {
  return (
    <GamsWidget
      WidgetComponent={{ Component: TripleForm }}
      WidgetDefProvider={{ Component: WidgetDefProvider, props: { globalPropName:"__gamsWidget__TripleForm", datastream: "WIDGET_TRIPLE_FORM" } }}
      WidgetDataProvider={{ Component: DataProvider }}
    />
    )

}


export default TripleFormWidget;

