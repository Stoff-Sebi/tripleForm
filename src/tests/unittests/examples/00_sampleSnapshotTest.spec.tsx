import jest from "jest";
import React from "react";
import renderer from 'react-test-renderer';
import App from "../../../components/App/TripleForm"

// NOT RECOMMENDED
// Example for a simple snapshot test
// WITHOUT USING ENYZME -> just Jest + Typescript

test('00_simpleSnapshotTest (no need for Enzyme): Snapshot tests are not recommended - Tests via Enzyme are better!', () => {
    // snapshots are created in first test run
    // then renderring is compared to previous generated snapshot
    // snapshot folder is mostly created in folder where testfiles are.
    const tree = renderer.create(<App/>).toJSON();
    expect(tree).toMatchSnapshot();
  });