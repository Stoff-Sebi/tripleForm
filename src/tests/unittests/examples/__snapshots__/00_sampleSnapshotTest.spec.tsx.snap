// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`00_simpleSnapshotTest (no need for Enzyme): Snapshot tests are not recommended - Tests via Enzyme are better! 1`] = `
<div
  className="App"
  data-test="component-app"
>
   
  <header
    className="App-header"
  >
    <h2>
      React - Web Komponente Demo
    </h2>
    <p>
      Dieser Paragraph wird von React gesteuert
    </p>
    <p>
      Und zwar von der Demo Komponente
    </p>
    <p
      className="paragraph"
    >
      Dieser Paragraph wird von einem css Modul gestyled
    </p>
    <p>
      Momentaner Server als environment variable: 
    </p>
    <a
      className="App-link"
      href="https://reactjs.org"
      rel="noopener noreferrer"
      target="_blank"
    >
      Learn React
    </a>
    <h2
      data-test="counter-display"
    >
      The counter is currently: 
      0
    </h2>
    <button
      data-test="increment-button"
      onClick={[Function]}
    >
      Increment Counter
    </button>
  </header>
</div>
`;
